File: Source/Rewards/OldSaves/BigNum.cs
public class BigNum
{
    public BigNumL List { get; set; }
    public int Base { get; set; }

    // Constructor
    public BigNum(int baseValue)
    {
        List = null;
        Base = baseValue;
    }

    // Clean method
    public void Clean()
    {
        BigNumL current = List;
        current?.Clean();
    }


    // DivSmall method
    public int DivSmall(int denom)
    {
        return List?.DivSmall(Base, denom) ?? 0;
    }

    // LastDigit method
    public int LastDigit()
    {
        BigNumL current = List;
        while (current?.Next != null)
        {
            current = current.Next;
        }
        return current?.Leaf ?? 0;
    }
}

public class BigNumL
{
    public int Leaf { get; set; }
    public BigNumL Next { get; set; }
    private static int _nalloc = 0;

    // Constructor
    public BigNumL()
    {
        Next = null;
        Leaf = 0;
        _nalloc++;
    }

    public static BigNumL Create()
    {
        return new BigNumL();
    }

    public bool Clean()
    {
        if (Next == null && Leaf == 0)
        {
            return true;
        }
        else if (Next != null && Next.Clean())
        {
            Next = null;
            return Leaf == 0;
        }
        else
        {
            return false;
        }
    }

    public int DivSmall(int baseValue, int denom)
    {
        int quotient = 0;
        int remainder = 0;
        int num;

        if (Next != null)
        {
            remainder = Next.DivSmall(baseValue, denom);
        }

        num = Leaf + remainder * baseValue;
        quotient = num / denom;
        remainder = num - quotient * denom;
        Leaf = quotient;
        return remainder;
    }
}

File: src/Rewards/OldSaves/BigNum.ts
export class BigNum {
    public List: BigNumL
    public Base = 0

    // Constructor
    public constructor(baseValue: number) {
        this.List = null as never
        this.Base = baseValue
    }

    // Clean method
    public Clean = () => {
        const current: BigNumL = this.List
        current?.Clean()
    }

    // DivSmall method
    public DivSmall = (denom: number) => {
        return this.List?.DivSmall(this.Base, denom) ?? 0
    }

    // LastDigit method
    public LastDigit(): number {
        let current: BigNumL = this.List
        while (!!current?.Next) {
            current = current.Next
        }
        return current?.Leaf ?? 0
    }
}

export class BigNumL {
    public Leaf = 0
    public Next: BigNumL
    private static _nalloc = 0

    // Constructor
    public constructor() {
        this.Next = null as never
        this.Leaf = 0
        BigNumL._nalloc++
    }

    public static Create(): BigNumL {
        return new BigNumL()
    }

    public Clean(): boolean {
        if (!this.Next && this.Leaf === 0) {
            return true
        } else if (this.Next && this.Next.Clean()) {
            this.Next = null as never
            return this.Leaf === 0
        } else {
            return false
        }
    }

    public DivSmall = (baseValue: number, denom: number) => {
        let remainder = 0

        if (!!this.Next) {
            remainder = this.Next.DivSmall(baseValue, denom)
        }

        const num = this.Leaf + remainder * baseValue
        const quotient: number = num / denom
        remainder = num - quotient * denom
        this.Leaf = quotient
        return remainder
    }
}
